### PR TITLE
fixes #1456 by adding rate limiter to shutdown queue

### LINKED ISSUE #1456: Prevent catastrophic tserver shutdown by restricting the status shutdown
We had a situation where the master had a bad hosts file for a 2 minute period. During that time, the master marked ~70% of the tservers as bad. When the good hosts file was then put in place, the master issued shutdowns to that 70% due to the master having failed the max number of status requests.

There should be a threshold percentage of tservers where the master refuses to issue that shutdown. If the set of tservers to be shutdown is greater than or equal to 10% as a default, do not issue the shutdowns. Clearly log an error level message so that users can monitor for this situation, but don't self destruct.

### PR DESCRIPTION
**Pull request by alerman**

### CODE REVIEW

File: server/master/src/main/java/org/apache/accumulo/master/Master.java [hunk 1]
```diff
@@ -1202,7 +1205,9 @@ public void run() {
               try {
                 TServerConnection connection = tserverSet.getConnection(server);
                 if (connection != null) {
-                  connection.halt(masterLock);
+                  if (rateLimiter.tryAcquire()) {
```
Comment (keith-turner) [inline server/master/src/main/java/org/apache/accumulo/master/Master.java]: If its not acquired, will the server still be removed from badServers?  I am not quite sure trying to match the curly braces looking at the diff.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java [hunk 2]
```diff
@@ -221,6 +222,7 @@
   volatile SortedMap<TServerInstance,TabletServerStatus> tserverStatus =
       Collections.unmodifiableSortedMap(new TreeMap<TServerInstance,TabletServerStatus>());
   final ServerBulkImportStatus bulkImportStatus = new ServerBulkImportStatus();
+  private RateLimiter shutdownServerRateLimiter = null;
```
Comment (keith-turner) [inline server/master/src/main/java/org/apache/accumulo/master/Master.java]: I don't see an assignment to this.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java (hunk 1 above)
Comment (alerman) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: @keith-turner  Yes, That isn't a change to what happens now. If it fails to issue the shutdown, the sever is removed and status is attempted three more times

File: server/master/src/main/java/org/apache/accumulo/master/Master.java (hunk 2 above)
Comment (alerman) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: good catch

File: server/master/src/main/java/org/apache/accumulo/master/Master.java (hunk 1 above)
Comment (keith-turner) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: > That isn't a change to what happens now. If it fails to issue the shutdown, the sever is removed and status is attempted three more times

Oh I see.  That is kinda cool, I really like that behavior.  I think a comment stating that this behavior is intentional would be good.  That if the rate limit does not allow a halt, then we want to reset the counter.

**Comment by keith-turner** [comment]
We could not add the new property and only add the rate limiter with a hard coded rate.  This bad server handling has a count that is not currently configurable.  So adding the rate limiter w/o a configuration option would just add one more constants to the bad server handling code.  The reason I am thinking this is that I am trying to determine what positive value the bad server handling code is, it may be that if we can not determine what the benefit is that we should remove the code.  If we add a configuration knob know, it would make that decision more complex.  I am not sure this course of action should be taken, just something I was thinking about and wanted to post if for discussion.

**Comment by alerman** [comment]
I agree with @keith-turner. I'm not sure this code shouldn't be removed. I removed the property in the latest commit. Personally, I think this should be a situation where we log and take no action. I could be convinced otherwise but at this time I can't come up with a reason that we should keep it.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java [hunk 3]
```diff
@@ -1202,7 +1205,13 @@ public void run() {
               try {
                 TServerConnection connection = tserverSet.getConnection(server);
                 if (connection != null) {
-                  connection.halt(masterLock);
+                  // Attempt to shutdown server only if able to acquire. If unable, this table
+                  // server
```
Comment (alerman) [inline server/master/src/main/java/org/apache/accumulo/master/Master.java]: Just realized a spelling error and the formatting is off. I will address this tomorrow.

**Comment by ctubbsii** [comment]
@alerman, what do you think about making the changes only for the next version of Accumulo (2.1, which is under development)? I see your pull request is against the 1.9 branch. Adding new features or behavior changes to an older stable release is something that makes me a bit nervous and I feel reluctant to encourage or support, unless there's a strong argument for it. I'd feel more comfortable restricting changes to bug fixes only... and not behavior changes or feature adds, even if they are internal and not user facing... unless there's a very good argument. What do you think?

**Comment by alerman** [comment]
@ctubbsii I agree that new features shouldn't necessarily end up in 1.9, but in this case, I believe this change in behavior is required. In the event that the master has an intermittent network issue (something we have seen on other nodes often) the entire cluster could be taken out. That seems like a behavior that we can't allow to remain as-is. Especially considering that this branch may be the basis for the first LTS release (if that is still planned).

I am perfectly fine with taking the discussion of whether to remove this behavior entirely to 2.x, but as it stands I believe this is a necessary safeguard to currently deployed 1.9 releases.

**Comment by keith-turner** [comment]
I think the behavior @alerman saw was undesirable and I would call it buggy.  I think this change w/o the property may improve the situation and qualify as a bug fix.  I do share the concern about this change causing new bugs, but that can be mitigated with targeted testing of the change and code review.

**Comment by keith-turner** [comment]
>  I could be convinced otherwise but at this time I can't come up with a reason that we should keep it.

I think there were legitimate use cases that this code was addressing.  Need to try to enumerate those if possible.  I opened #1460

**Comment by milleruntime** [comment]
I think we should stop the halt of tservers for 1.9 and just spam the log that the tserver is not responding.  The change in behavior would be stopping a dangerous behavior.  Similar to the check we recently added to prevent dangerous behavior of cloning of the metadata table.   We should only change a few lines of code for 1.9 and then refactor in 2.1.

**Comment by milleruntime** [comment]
> I think we should stop the halt of tservers for 1.9 and just spam the log that the tserver is not responding. The change in behavior would be stopping a dangerous behavior. Similar to the check we recently added to prevent dangerous behavior of cloning of the metadata table. We should only change a few lines of code for 1.9 and then refactor in 2.1.

This would be the easiest to maintain but I was unaware an unresponsive tserver could be so catastrophic for a cluster.  In this case, what you have is probably the best for 1.9.

**Comment by keith-turner** [comment]
@alerman I merged this change into master in commit bf3d3d4c3cc12c4c1d016d7003dffc185a7dc057  It was tricky to merge, if you have a second could you take a look to make sure I got it right?

While merging it I noticed the rate limiter is created in the method.  I was thinking the rate limiter existed as an instance variable not a local variable.

**Pull request merged**